### PR TITLE
GSS BZ-2214838 Automation, Test Rook Reclaim Namespace

### PR DESCRIPTION
**Pull request by OdedViner**
https://bugzilla.redhat.com/show_bug.cgi?id=2214838#c51

### CODE REVIEW

File: tests/manage/z_cluster/test_rook_reclaim_namespace.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            timeout=100,
+            sleep=5,
+            func=verify_log_exist_in_pods_logs,
+            pod_names=pod_names,
+            expected_log=expected_log,
+        )
+        if not sample.wait_for_func_status(result=True):
+            raise ValueError(
+                f"The expected log '{expected_log}' does not exist in {pod_names} pods"
+            )
```
Comment (OdedViner) [inline tests/manage/z_cluster/test_rook_reclaim_namespace.py:90]: $ oc logs csi-rbdplugin-5vzdn --since=60s

File: tests/functional/z_cluster/test_rook_reclaim_namespace.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+
+@brown_squad
+@tier2
+@ignore_leftovers
+@bugzilla("2214838")
+@skipif_external_mode
+@skipif_ocs_version("<4.13")
+@pytest.mark.polarion_id("OCS-XXXX")
+class TestRookReclaimNamespace(ManageTest):
```
[the remark this answers is not in the document]
Comment (OdedViner) [reply tests/functional/z_cluster/test_rook_reclaim_namespace.py:34]: done

File: tests/functional/z_cluster/test_rook_reclaim_namespace.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+)
+
+log = logging.getLogger(__name__)
+
+
+@brown_squad
+@tier2
+@ignore_leftovers
+@bugzilla("2214838")
+@skipif_external_mode
```
[the remark this answers is not in the document]
Comment (OdedViner) [reply tests/functional/z_cluster/test_rook_reclaim_namespace.py:31]: I did not check, 
Deleted  "@skipif_external_mode"

File: tests/functional/z_cluster/test_rook_reclaim_namespace.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        if not sample.wait_for_func_status(result=True):
+            raise ValueError(
+                f"The expected log '{expected_log}' does not exist in {pod_names} pods"
+            )
+
+        log.info(f"Delete pod {pod_obj.name}")
+        pod_obj.delete()
+
+        log.info(f"Delete reclaimspacejob CR {self.reclaim_space_job_obj_1.name}")
+        self.reclaim_space_job_obj_1.delete()
```
[the remark this answers is not in the document]
Comment (OdedViner) [reply tests/functional/z_cluster/test_rook_reclaim_namespace.py:112]: done

File: tests/functional/z_cluster/test_rook_reclaim_namespace.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        log.info("Recreate reclaimspacejob CR")
+        self.reclaim_space_job_obj_2 = create_reclaim_space_job(pvc_name=pvc_obj.name)
```
[the remark this answers is not in the document]
Comment (OdedViner) [reply tests/functional/z_cluster/test_rook_reclaim_namespace.py:115]: done

File: tests/functional/z_cluster/test_rook_reclaim_namespace.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+        log.info(f"Sleep 70 seconds so the logs in {pod_names} will be updated")
```
[the remark this answers is not in the document]
Comment (OdedViner) [reply tests/functional/z_cluster/test_rook_reclaim_namespace.py:117]: @vkathole What do you think?

File: tests/functional/z_cluster/test_rook_reclaim_namespace.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        pod_obj = pod_factory(
+            interface=constants.CEPHFILESYSTEM,
+            pvc=pvc_obj,
+            status=constants.STATUS_RUNNING,
+            pod_dict_path=pod_dict_path,
+            raw_block_pv=raw_block_pv,
+        )
+
+        log.info(f"Create reclaimspacejob CR to run on pvc {pvc_obj.name}")
+        self.reclaim_space_job_obj_1 = create_reclaim_space_job(pvc_name=pvc_obj.name)
```
[the remark this answers is not in the document]
Comment (OdedViner) [reply tests/functional/z_cluster/test_rook_reclaim_namespace.py:89]: I created generator to verify log exist [timeout=100sec]

File: tests/functional/z_cluster/test_rook_reclaim_namespace.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        expected_log = "skipping sparsify operation"
+        pod_names = get_pod_name_by_pattern(pattern="csi-rbdplugin-provisioner")
```
[the remark this answers is not in the document]
Comment (OdedViner) [reply tests/functional/z_cluster/test_rook_reclaim_namespace.py]: done

File: tests/manage/z_cluster/test_rook_reclaim_namespace.py (hunk 1 above)
Comment (OdedViner) [reply tests/manage/z_cluster/test_rook_reclaim_namespace.py]: done

File: tests/functional/z_cluster/test_rook_reclaim_namespace.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            size=2,
+        )
+        pod_dict_path = constants.NGINX_POD_YAML
+        raw_block_pv = False
+
+        log.info(
+            f"Created new pod sc_name={constants.CEPHFILESYSTEM} size=10Gi, "
+            f"access_mode={constants.ACCESS_MODE_RWX}, volume_mode={constants.VOLUME_MODE_FILESYSTEM}"
+        )
+        pod_obj = pod_factory(
```
[the remark this answers is not in the document]
Comment (OdedViner) [reply tests/functional/z_cluster/test_rook_reclaim_namespace.py]: on step 7 I delete only the pod

File: ocs_ci/ocs/resources/pod.py [hunk 10]
```diff
@@ -1755,6 +1756,7 @@ def get_pod_logs(
     namespace (str): Namespace of the pod
     previous (bool): True, if pod previous log required. False otherwise.
     all_containers (bool): fetch logs from all containers of the resource
+    since (str): only return logs newer than a relative duration like 5s, 2m, or 3h.
```
Comment (PrasadDesala) [inline ocs_ci/ocs/resources/pod.py:1759]: same as my above comment

File: tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    ManageTest,
+    tier2,
+    bugzilla,
+    skipif_ocs_version,
+)
+
+log = logging.getLogger(__name__)
+
+
+@brown_squad
```
Comment (PrasadDesala) [inline tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py:25]: Since this test is related to pvcs, should be be Greensquad here?

File: tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+@tier2
+@bugzilla("2214838")
+@skipif_ocs_version("<4.13")
+@pytest.mark.polarion_id("OCS-OCS-5424")
```
Comment (PrasadDesala) [inline tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py:29]: correct the polarion id, ocs is added twice

File: tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        time.sleep(70)
+
+        log.info("Verify logs does not show 'skipping sparsify operation' message.")
+        log_exist = verify_log_exist_in_pods_logs(
+            pod_names=pod_names, expected_log=expected_log, since="70s"
+        )
+        if log_exist:
+            raise ValueError(
+                f"The expected log '{expected_log}' exist in {pod_names} pods after reclaimspacejob deletion"
+            )
```
Comment (PrasadDesala) [inline tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py:127]: can we call the function using Timeoutsampler? that way we done need to hardcode the sleep time

File: tests/functional/z_cluster/test_rook_reclaim_namespace.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (PrasadDesala) [reply tests/functional/z_cluster/test_rook_reclaim_namespace.py]: this comment seems not incorporated @OdedViner

File: ocs_ci/helpers/helpers.py [hunk 14]
```diff
@@ -4449,3 +4449,38 @@ def check_selinux_relabeling(pod_obj):
...
+    Args:
+        pod_names (list): Name of the pod
+        expected_log:
+        container (str): Name of the container
+        namespace (str): Namespace of the pod
+        all_containers_flag (bool): fetch logs from all containers of the resource
+        since (str): only return logs newer than a relative duration like 5s, 2m, or 3h.
+    Returns:
+        bool: return True if log exist otherwise False
+    """
```
[the remark this answers is not in the document]
Comment (OdedViner) [reply ocs_ci/helpers/helpers.py:4474]: done

File: ocs_ci/helpers/helpers.py [hunk 15]
```diff
@@ -4449,3 +4449,38 @@ def check_selinux_relabeling(pod_obj):
...
+    Verify log exist in pods logs.
+
+    Args:
+        pod_names (list): Name of the pod
+        expected_log:
+        container (str): Name of the container
+        namespace (str): Namespace of the pod
+        all_containers_flag (bool): fetch logs from all containers of the resource
+        since (str): only return logs newer than a relative duration like 5s, 2m, or 3h.
```
[the remark this answers is not in the document]
Comment (OdedViner) [reply ocs_ci/helpers/helpers.py:4471]: no, because we need to mention the s[second] m[minutes] and etc...

File: ocs_ci/ocs/resources/pod.py (hunk 10 above)
Comment (OdedViner) [reply ocs_ci/ocs/resources/pod.py:1759]: no, because we need to mention the s[second] m[minutes] and etc...

File: tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py (hunk 11 above)
Comment (OdedViner) [reply tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py:25]: done

File: tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py (hunk 12 above)
Comment (OdedViner) [reply tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py:29]: done

File: tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py (hunk 13 above)
Comment (OdedViner) [reply tests/functional/pv/space_reclaim/test_rook_reclaim_namespace.py:127]: it is more difficult because  I need to change the "since" flag each iteration.. In my process , we wait 70 seconds and then verify log does not exist

File: tests/functional/z_cluster/test_rook_reclaim_namespace.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (OdedViner) [reply tests/functional/z_cluster/test_rook_reclaim_namespace.py]: done

File: ocs_ci/helpers/helpers.py [hunk 16]
```diff
@@ -4449,3 +4449,38 @@ def check_selinux_relabeling(pod_obj):
...
+    Returns:
```
[the remark this answers is not in the document]
Comment (OdedViner) [reply ocs_ci/helpers/helpers.py:4473]: done

**Review by PrasadDesala — APPROVED** [review]

File: ocs_ci/helpers/helpers.py [hunk 17]
```diff
@@ -4451,32 +4451,38 @@ def check_selinux_relabeling(pod_obj):
...
 ):
     """
-    Verify log exist in pod logs.
+    Verify log exist in pods logs.
 
     Args:
-        pod_name (str): Name of the pod
-        expected_log (str): the expected logs in "oc logs" command
+        pod_names (list): Name of the pod
+        expected_log:
```
Comment (yitzhak12) [inline ocs_ci/helpers/helpers.py:4467]: Why did you remove the `(str): the expected logs in "oc logs" command` from the docstring?

**Review by yitzhak12 — DISMISSED** [review]

File: ocs_ci/helpers/helpers.py (hunk 17 above)
Comment (OdedViner) [reply ocs_ci/helpers/helpers.py]: fixed

**Review by yitzhak12 — APPROVED** [review]

**Review by PrasadDesala — APPROVED** [review]
